File: Python/CIP/04 M5 - Leetcode 560 - Subarray Sum Equals K/Leetcode 560 - Subarray Sum Equals K/O(n)/560-external.js

// https://leetcode.com/problems/subarray-sum-equals-k/discuss/1138185/JavaScript-Hashmap

const subarraySum = (nums, k) => {
    const hashmap = {0: 1}
    let count = 0, total = 0
    for(const num of nums) {
        total += num
        if(hashmap[total - k]) count += hashmap[total - k]
        hashmap[total] ? hashmap[total]++ : hashmap[total] = 1
    }
    return count
};

File: Python/CIP/04 M5 - Leetcode 560 - Subarray Sum Equals K/Leetcode 560 - Subarray Sum Equals K/O(n)/560-external.py

# https://leetcode.com/problems/subarray-sum-equals-k/discuss/592003/C%2B%2BPython-solution-O(n)

def subarraySum(self, nums: List[int], k: int) -> int:
	ret = 0
	sum_freq = defaultdict(int)
	cur_sum = 0
	for num in nums:
		cur_sum += num
		if cur_sum == k:
			ret += 1
		if cur_sum - k in sum_freq:
			ret += sum_freq.get(cur_sum - k) 
		sum_freq[cur_sum] += 1
	return ret
